nx/lib/nx/binary_backend.ex

```diff
@@ -1679,6 +1679,32 @@ defmodule Nx.BinaryBackend do
     from_binary(out, data)
   end
 
+  @impl true
+  def take(out, tensor, indices, axis) do
+    %T{type: {_, idx_size}} = indices
+
+    # We iterate over the indices in a flat manner,
+    # and take a unit tensor slice along axis given
+    # by each index. Then we concatenate the tensors
+    # along the axis, which gives us the result with
+    # index dimensions flattened and we just reshape.
+
+    tensor_rank = tuple_size(tensor.shape)
+    slice_start = List.duplicate(0, tensor_rank)
+    slice_lengths = tensor.shape |> Tuple.to_list() |> List.replace_at(axis, 1)
+
+    slices =
+      for <<bin::size(idx_size)-bitstring <- to_binary(indices)>> do
+        idx = binary_to_number(bin, indices.type)
+        slice_start = List.replace_at(slice_start, axis, idx)
+        Nx.slice(tensor, slice_start, slice_lengths)
+      end
+
+    slices
+    |> Nx.concatenate(axis: axis)
+    |> Nx.reshape(out.shape, names: out.names)
```

Comment:
@josevalim I might be recalling this wrong, but I remember there being an issue with calling `Nx`-functions from binary backend. Is this the case?

@jonatanklosko if this is the case, you could just use `concatenate` directly from the binary backend (the signature is just below on line 1709).

Perhaps you could extract the `for` expression it uses to a defp, so you can avoid converting to and from binaries all the time

Reply:
I was wondering about the usage of `Nx` functions, but we already use `Nx.transpose` and `Nx.pad` quite a lot in `BinaryBackend`, so I assumed it's fine :thinking: @josevalim?

> Perhaps you could extract the for expression it uses to a defp, so you can avoid converting to and from binaries all the time

The for loop in `take` goes over the indices tensor to produce a list of input tensor slices, then `concat` goes over these input tensor slices and combines them. These seem like two separate things, I may be missing something though :)

Reply:
You are correct, I think we changed to not call Nx from it but I don’t remember the reason. This is also related to having optional functions per backend. For example, we shouldn’t need to grad it if it is implemented as part of slice and friends. But likely there is a faster grad for it too.

Reply:
> The for loop in take goes over the indices tensor to produce a list of input tensor slices, then concat goes over these input tensor slices and combines them. These seem like two separate things, I may be missing something though :)

Making myself more clear, I was refering to the `for` in `concatenate`. If it's extracted to a defp, it would be easier to call `do_concatenate` with the binary representation directly, no need to build a tensor just to "unbuild" it  afterwards

Reply:
Hmm, in order to call the backend implementations of `slice` and `concatenate` we would need to compute output shape, names and other arguments, effectively duplicating what we already do in the `Nx` calls.